omnigibson/utils/bddl_utils.py

```diff
@@ -89,9 +127,810 @@ def process_single_condition(condition):
+            if condition is None:
+                continue
+
+            # Sampled conditions must always be positive
+            # Non-positive (e.g.: NOT onTop) is not restrictive enough for sampling
+            if condition.STATE_NAME in KINEMATIC_STATES_BDDL and not positive:
+                return "Initial condition has negative kinematic conditions: {}".format(condition.body)
+
+            # Infer the group the condition and its object instances belong to
+            # (a) Kinematic (binary) conditions, where (ent0, ent1) are both objects
+            # (b) Particle (binary) conditions, where (ent0, ent1) are (object, substance)
+            # (c) Future conditions, where (ent0,) can be either an objects or substance
+            # (d) Unary conditions, where (ent0,) is an object
+            # Binary conditions have length 2: (ent0, ent1)
+            if len(condition.body) == 2:
+                group = "particle" if condition.body[1] in self._substance_instances else "kinematic"
+            else:
+                assert len(condition.body) == 1, \
+                    f"Got invalid parsed initial condition; body length should either be 2 or 1. " \
+                    f"Got body: {condition.body} for condition: {condition}"
+                group = "future" if isinstance(condition, ObjectStateFuturePredicate) else "unary"
+            sampling_groups[group].append(condition.body)
+            self._object_sampling_conditions[group].append((condition, positive))
+
+            # If the condition involves any non-sampleable object (e.g.: furniture), it's a non-sampleable condition
+            # This means that there's no ordering constraint in terms of sampling, because we know the, e.g., furniture
+            # object already exists in the scene and is placed, so these specific conditions can be sampled without
+            # any dependencies
+            if len(self._non_sampleable_object_instances.intersection(set(condition.body))) > 0:
+                self._non_sampleable_object_conditions.append((condition, positive))
+
+        # Now, sort each group, ignoring the futures (since they don't get sampled)
+        # First handle kinematics, then particles, then unary
+
+        # Start with the non-sampleable objects as the first sampled set, then infer recursively
+        cur_batch = self._non_sampleable_object_instances
+        while len(cur_batch) > 0:
+            next_batch = set()
+            for condition, _ in self._object_sampling_conditions["kinematic"]:
+                if condition.body[1] in cur_batch:
+                    next_batch.add(condition.body[0])
+            cur_batch = next_batch
+            self._object_sampling_orders["kinematic"].append(cur_batch)
+        # pop final value since it's an empty set
+        self._object_sampling_orders["kinematic"].pop(-1)
+
+        # Now parse particles -- this time, in reverse order, starting from the final kinematic group
```

Comment:
why reverse order?

Reply:
i think this was based on previous discussion -- imagine two nested containers (cup inside pot). Maybe we should sample the cup's particles first before the pot's. But tbh, it might not matter the ordering.

Reply:
add a comment

Reply:
done